onnxruntime/core/providers/cuda/tensor/concat_impl.cu

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/cuda_common.h"
+#include "concat_impl.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T>
+__global__ void _ConcatKernel(const fast_divmod block_size_including_axis_dim_div,
+                              const fast_divmod block_size_inside_axis_dim_div,
+                              const int64_t* concat_sizes,
+                              const int64_t* concat_sizes_range,
+                              const int num_inputs,
+                              T* output_data,
+                              const void** input_ptr,
+                              const CUDA_LONG N) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
+  CUDA_LONG input_pos = 0;
+
+  int outter_block_index = 0;
+  int block_index = 0;
+  int offset = 0;
+
+  int input_index = 0;
+  int block_offset = 0;
+
+  block_size_including_axis_dim_div.divmod(id, outter_block_index, offset);
+  block_size_inside_axis_dim_div.divmod(offset, block_index, offset);
+
+  for (int i = 0; i < num_inputs; ++i) {
+    int64_t range_left = (i == 0) ? 0 : concat_sizes_range[i - 1];
+    if ((range_left <= block_index) && (block_index < concat_sizes_range[i])) {
+      input_index = i;
+      block_offset = block_index - range_left;
+      break;
+    }
+  }
+
+  input_pos = (outter_block_index * concat_sizes[input_index] + block_offset) *
+               block_size_inside_axis_dim_div.d_ +
+               offset;
+
+  output_data[id] = reinterpret_cast<const T*>(input_ptr[input_index])[input_pos];
```

Comment:
It seem like every thread writes into output_data[id] only 1 time -- so 1 thread implies 1 output index is populated. If the output tensor is large (i.e. larger than number of threads in the grid), how are the remaining indexes being populated?

Reply:
there's also a blocksPerGrid, check line 52

---
In reply to: [299716784](https://github.com/microsoft/onnxruntime/pull/1333#discussion_r299716784) [](ancestors = 299716784)